src/editor/dag.rs

```diff
@@ -177,6 +177,23 @@ impl std::fmt::Display for Insertable {
     }
 }
 
+/// A datastructure that stores the current tree before performing some editing.
+struct Snapshot<'arena, Node: Ast<'arena>> {
+    root: &'arena Node,
+    cursor_before: Path,
+    cursor_after: Path,
```

Comment:
Very small nitpick, but I think `cursor_before`, `root`, `cursor_after` is a more intuitive order.